### PR TITLE
chore: cypress snapshot types

### PR DESCRIPTION
**Pull request by KelvinOm**
## Description
1. Add scripts for local e2e testing in the docker container. 
2. Add types for cypress-image-snapshot

In this PR, I also added the following PRS as they affect screenshot tests as well. I did this to speed up the process and unblock the team. 

#34528
#34546
#34676
#34729
#34638
#34639
#34511

To run E2E tests locally in docker, you need to do the following:
1. Run FE locally and prepare the tests for local launch. See the instructions [here](https://github.com/appsmithorg/appsmith/blob/release/contributions/ClientSetup.md). 
2. Run `yarn cypress:snapshot:docker:build` — this will create a docker container with the necessary environment.
3. Run `yarn cypress:snapshot:docker "./cypress/e2e/Regression/ClientSide/Anvil/Widgets/*_spec.ts" updateSnapshots=false`. Here we can use the path to a specific file, or set `updateSnapshots=true` flag to update the screenshots. 

## Automation

/ok-to-test tags="@tag.Anvil"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/9844579277>
> Commit: 75f26599b149f831051fbd631aaa059084b226dd
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=9844579277&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Anvil`
> Spec:
> <hr>Mon, 08 Jul 2024 18:37:36 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [x] No

## Summary by CodeRabbit

- **New Features**
  - Added support for image snapshot testing with the `@types/cypress-image-snapshot` dependency.

- **Refactor**
  - Updated test specifications by removing unnecessary `triggerInputInvalidState()` calls.
  - Reorganized and improved efficiency of image snapshot methods for various devices.

- **Chores**
  - Updated `Dockerfile` to configure the Cypress environment with specific versions for dependencies.
  - Changed import paths in `e2e.js` for better module resolution. 

- **Style**
  - Fixed a comment typo in Cypress plugin configuration.

### CODE REVIEW

File: app/client/cypress/e2e/Regression/ClientSide/Anvil/Widgets/AnvilIconButtonWidgetSnapshot_spec.ts [hunk 1]
```diff
@@ -13,6 +13,7 @@ describe(
     });
 
     it("1. Canvas Mode", () => {
+      cy.wait(5000);
```
Comment (ApekshaBhosale) [inline app/client/cypress/e2e/Regression/ClientSide/Anvil/Widgets/AnvilIconButtonWidgetSnapshot_spec.ts:16]: @sagar-qa007 can you check why cypress didn't suggest anything here?

**Review by ApekshaBhosale — CHANGES_REQUESTED** [review]

**Comment by sagar-qa007** [comment]
@coderabbitai full review

File: app/client/cypress/e2e/Regression/ClientSide/Anvil/Widgets/AnvilIconButtonWidgetSnapshot_spec.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (sagar-qa007) [reply app/client/cypress/e2e/Regression/ClientSide/Anvil/Widgets/AnvilIconButtonWidgetSnapshot_spec.ts:16]: @ApekshaBhosale I have rerun full review and coderabbit gave us comment.

File: app/client/cypress/snapshots/AnvilButtonWidgetSnapshot_spec.ts/anvilButtonWidgetDeployiphone-6.snap.png
Comment (jsartisan) [inline app/client/cypress/snapshots/AnvilButtonWidgetSnapshot_spec.ts/anvilButtonWidgetDeployiphone-6.snap.png:1]: When we change the locator, we get snapshots like this :/ not sure if they are reliable.

File: app/client/cypress/snapshots/AnvilCheckboxGroupWidgetSnapshot_spec.ts/anvilCheckboxGroupWidgetDeployiphone-6.snap.png
Comment (jsartisan) [inline app/client/cypress/snapshots/AnvilCheckboxGroupWidgetSnapshot_spec.ts/anvilCheckboxGroupWidgetDeployiphone-6.snap.png:1]: Same here. Unreliable screenshot due to change of locator.

File: app/client/cypress/snapshots/AnvilCheckboxWidgetSnapshot_spec.ts/anvilCheckboxWidgetDeployiphone-6.snap.png
Comment (jsartisan) [inline app/client/cypress/snapshots/AnvilCheckboxWidgetSnapshot_spec.ts/anvilCheckboxWidgetDeployiphone-6.snap.png:1]: At some places, it pure transparent and somewhere it is gray background.

Comment (KelvinOm) [reply app/client/cypress/snapshots/AnvilCheckboxWidgetSnapshot_spec.ts/anvilCheckboxWidgetDeployiphone-6.snap.png:1]: I can't say that these screenshots are unreliable. They are like that due to the specifics of taking screenshots by cypress. We can set the height of the images (make it the same as the height of the viewport), but I would not like to do this now. If there are any difficulties with these screenshots, we will change them later.

**Review by jsartisan — APPROVED** [review]

**Comment by ApekshaBhosale** [comment]
@KelvinOm LGTM

File: app/client/cypress/e2e/Regression/ClientSide/Anvil/Widgets/AnvilCheckboxGroupWidgetSnapshot_spec.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+describe(
+  `${ANVIL_EDITOR_TEST}: Anvil tests for Checkbox Group Widget`,
+  { tags: ["@tag.Anvil"] },
+  () => {
+    before(() => {
+      agHelper.AddDsl("anvilCheckboxGroupWidget");
+    });
+
+    it("1. Canvas Mode", () => {
```
Comment (sagar-qa007) [inline app/client/cypress/e2e/Regression/ClientSide/Anvil/Widgets/AnvilCheckboxGroupWidgetSnapshot_spec.ts:15]: Can we also have E2E functional test cases here?

Comment (KelvinOm) [reply app/client/cypress/e2e/Regression/ClientSide/Anvil/Widgets/AnvilCheckboxGroupWidgetSnapshot_spec.ts:15]: We do not plan to do this. Read more [here](https://www.notion.so/appsmith/Anvil-E2E-testing-1206f4702bb24fccbd23ddd1b4de1361). 

We will cover dnd, selection and highlighting with functional tests. Everything else is just screenshots and we won't mix all this.

Comment (sagar-qa007) [reply app/client/cypress/e2e/Regression/ClientSide/Anvil/Widgets/AnvilCheckboxGroupWidgetSnapshot_spec.ts:15]: Okay sure.

**Review by sagar-qa007 — APPROVED** [review]

**Pull request merged**